### PR TITLE
[Datasets] Refactor Ray Data API documentation

### LINKED ISSUE #30692: [Datasets] Make Ray Data API reference page more easier to read
### Description

Current we have an Index Page - https://docs.ray.io/en/latest/data/api/api.html, and Individual API Page - https://docs.ray.io/en/latest/data/api/dataset.html#basic-transformations. It has two problems:

* The API Page has all APIs, which is crowded, hard to read, and pass around link to reference individual API. It also does not work well with search engine. E.g. Googling `ray data map_batches` does not show up the API page for `map_batches`. But googling `pyspark map` shows up the API page in first seveal results.
* Summary Page for a group of related APIs is hard to discover. Currently when user clicks "[Basic Transformations](https://docs.ray.io/en/latest/data/api/dataset.html#basic-transformations)" in our Index page - https://docs.ray.io/en/latest/data/api/api.html, it directs users to a page with all APIs -  https://docs.ray.io/en/latest/data/api/dataset.html#basic-transformations . We should have a Summary Page to list all APIs of `Basic Transformations`, and one-line explanation of those APIs.

It would be better to have flow as:

Index Page -> Summary Page -> Individual API Page.

1. Index page gives an index of all APIs based on category (current is good, no need to change)
2. Index page directs user to Summary page.
3. Summary page directs user to Individual API page (each API has its own page, not put them together).

E.g.
Index page: https://spark.apache.org/docs/latest/api/python/reference/index.html
Summary page: https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/series.html
Individual API page: https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/api/pyspark.pandas.Series.map.html#pyspark.pandas.Series.map

### Link

https://docs.ray.io/en/latest/data/api/api.html

### PR DESCRIPTION
**Pull request by c21**
Signed-off-by: Cheng Su <scnju13@gmail.com>

## Why are these changes needed?

Please review the new documentation in https://ray--31204.org.readthedocs.build/en/31204/data/api/api.html .

This PR is to refactor Ray Data API documentation as specified in https://github.com/ray-project/ray/issues/30692 . The motivation is to have a better layout and view of all API references (same as other systems - [Pandas](https://pandas.pydata.org/docs/reference/index.html), [NumPy](https://numpy.org/doc/stable/reference/index.html) and [Spark](https://spark.apache.org/docs/latest/api/python/reference/index.html)). 

The new structure has 3 pages:

1.index page (same as before, [link](https://ray--31204.org.readthedocs.build/en/31204/data/api/api.html))
2.summary page for each group of APIs (newly added, [link](https://ray--31204.org.readthedocs.build/en/31204/data/api/dataset.html))
3.doc page for individual APIs (newly added, [link](https://ray--31204.org.readthedocs.build/en/31204/data/api/doc/ray.data.Dataset.map_batches.html))

In addition, the side bar also has list of APIs per category:

Screen Shot 2022-12-20 at 8 49 54 PM

## Related issue number

Closes https://github.com/ray-project/ray/issues/30692

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/_static/css/custom.css [hunk 1]
```diff
@@ -245,8 +245,13 @@ th.tune-colab {
...
 .table td {
-  /* Remove row spacing */
-  padding: 0;
+  /* Remove row spacing on the left */
+  padding-left: 0;
+}
+
+.table thead th {
+  /* Remove row spacing on the left */
+  padding-left: 0;
```
Comment (c21) [inline doc/source/_static/css/custom.css:254]: This is required to make the summary page readable. Thanks @sijieamoy for suggesting this fix!

Before this change, the summary table looks bad as below (characters are packed together):

Screen Shot 2022-12-20 at 11 59 13 AM

File: doc/source/data/api/grouped_dataset.rst [hunk 2]
```diff
@@ -3,29 +3,54 @@
 GroupedDataset API
 ==================
 
-.. autoclass:: ray.data.grouped_dataset.GroupedDataset
-    :members:
+.. currentmodule:: ray.data
```
Comment (c21) [inline doc/source/data/api/grouped_dataset.rst:6]: This page is not only refactored, but also restructured.

Reference: 
Pandas groupby - https://pandas.pydata.org/docs/reference/groupby.html
PySpark groupby - https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/groupby.html

File: doc/source/data/api/data_representations.rst [hunk 3]
```diff
@@ -3,54 +3,53 @@
...
-.. autoclass:: ray.data.block.BlockExecStats
-    :members:
-
-.. autoclass:: ray.data.block.BlockMetadata
-    :members:
-
-.. autoclass:: ray.data.block.BlockAccessor
-    :members:
+.. autosummary::
+   :toctree: doc/
```
Comment (c21) [inline doc/source/data/api/data_representations.rst:14]: This `:toctree: doc/` forces the individual API to be generated in separate page. with `.../doc/<API-name>.html` link.

Example:

https://ray--31204.org.readthedocs.build/en/31204/data/api/doc/ray.data.Dataset.map_batches.html

**Review by jianoaix — COMMENTED** [review]
This looks nice and it's consistent with other major OSS data libraries, so LG for just data API itself.

I think we will also need to take care of the consistency with other Ray libraries. This will make data API docs different than the rest of Ray, so we may need to reach a consensus with the ray-docs group that this is the structure we want for Ray overall.

File: doc/source/data/api/dataset.rst [hunk 4]
```diff
@@ -3,289 +3,160 @@
...
-.. automethod:: ray.data.Dataset.drop_columns
-
-.. automethod:: ray.data.Dataset.select_columns
-
-.. automethod:: ray.data.Dataset.random_sample
-
-.. automethod:: ray.data.Dataset.limit
+.. autosummary::
+   :toctree: doc/
+   :nosignatures:
```
Comment (clarkzinzow) [inline doc/source/data/api/dataset.rst:22]: I feel like not having the function signatures really hurts understanding the one-liner API description, and the [Pandas](https://pandas.pydata.org/docs/reference/groupby.html) and [PySpark](https://spark.apache.org/docs/latest/api/python/reference/pyspark.pandas/groupby.html) docs _do_ include the signatures; can you say more about why you decided to not have the signatures displayed?

Comment (c21) [reply doc/source/data/api/dataset.rst:22]: I guess I am on the other side, that feels the signature is not that useful in one-liner description, given it's also truncated in UI. But I agree this is what other systems are doing. I can add it if this is what people want. WDYT? @jianoaix

**Comment by c21** [comment]
> I think we will also need to take care of the consistency with other Ray libraries. This will make data API docs different than the rest of Ray, so we may need to reach a consensus with the ray-docs group that this is the structure we want for Ray overall.

Yes this is fair. WDYT? @maxpumperla, @ericl and @zhe-thoughts, thanks.

File: doc/source/data/api/dataset.rst (hunk 4 above)
Comment (clarkzinzow) [reply doc/source/data/api/dataset.rst:22]: Got it, I get that perspective as well! My opinion is that, for most of the Datasets APIs, the first argument or two are typically much more important than the remaining arguments, and even with the signature being truncated, the names of the first few arguments adds a lot of valuable context.

For example:
```
read_parquet(paths[, filesystem, parallelism, ...])

map_batches(fn[, batch_size, batch_format, ...])
```

+1 on getting @jianoaix's perspective, we can do a quick 3-person majority rule here, if needed!

Comment (jianoaix) [reply doc/source/data/api/dataset.rst:22]: Yeah, I personally like signatures (and type annotations in general, although may not fit here) and they do convey useful information to the readers (especially when with good namings :).

File: doc/source/data/api/data_representations.rst [hunk 5]
```diff
@@ -3,54 +3,53 @@
 Data Representations
 ====================
```
Comment (ericl) [inline doc/source/data/api/data_representations.rst:4]: Aren't these internal APIs?

**Review by ericl — COMMENTED** [review]
Couple thoughts:

1. https://ray--31204.org.readthedocs.build/en/31204/data/api/doc/ray.data.Dataset.map_batches.html --- if you look at this, it's very hard to read. There are too many args, the type signature is way too verbose, and there are 5 different Tip blocks.

2. After this PR, every method now has a sidebar entry. This seems a bit much: could we hide these entries from the sidebar, or not use headings?

**Comment by c21** [comment]
>https://ray--31204.org.readthedocs.build/en/31204/data/api/doc/ray.data.Dataset.map_batches.html --- if you look at this, it's very hard to read. There are too many args, the type signature is way too verbose, and there are 5 different Tip blocks.

Yes, and [it's the same as current page](https://docs.ray.io/en/master/data/api/dataset.html#ray.data.Dataset.map_batches). This PR is just moving the doc of `map_batches` to a separate page. IMO the current page is even harder to read. We should definitely improve the doc of individual API for sure separately.

> After this PR, every method now has a sidebar entry. This seems a bit much: could we hide these entries from the sidebar, or not use headings?

I don't have a strong opinion on this. WDYT @clarkzinzow? Remember you think the sidebar is very useful.

File: doc/source/data/api/data_representations.rst (hunk 5 above)
Comment (c21) [reply doc/source/data/api/data_representations.rst:4]: > Aren't these internal APIs?

They are internal. I feel we should remove them (this PR is mostly refactoring stuff). cc @matthewdeng.

File: doc/source/data/api/dataset.rst (hunk 4 above)
Comment (c21) [reply doc/source/data/api/dataset.rst:22]: Cool, let me add signature then.

**Comment by clarkzinzow** [comment]
@ericl @c21 The sidebar entries are very useful for fast API doc scanning, I always use them for any libraries that have classes with many methods or modules with many functions, and is a somewhat common practice in the Python data library ecosystem:
- Pandas: https://pandas.pydata.org/docs/reference/frame.html
- NumPy: https://numpy.org/doc/1.24/reference/arrays.ndarray.html

This could be considered less of a must-have now that we've reduced the index pages for e.g. the `Dataset` class to be a table-list of single-line method summaries and link-outs.

File: doc/source/data/api/data_representations.rst (hunk 5 above)
Comment (matthewdeng) [reply doc/source/data/api/data_representations.rst]: Some of these APIs are Public or Developer APIs. cc @clarkzinzow

Comment (clarkzinzow) [reply doc/source/data/api/data_representations.rst]: The block/batch types are `DeveloperAPI`, the `TableRow` type is `PublicAPI`, and the tensor extension types are `PublicAPI`.

Comment (c21) [reply doc/source/data/api/data_representations.rst:4]: Cool, then let's keep this file.

**Comment by c21** [comment]
This PR is ready for review again. cc @ericl, @maxpumperla, @clarkzinzow, @jianoaix and @richardliaw. Thanks.

> After this PR, every method now has a sidebar entry. This seems a bit much: could we hide these entries from the sidebar, or not use headings?

@ericl - this seems to be default behavior, and I don't find an easy way to hide by default. One can hide manually by click the upper arrow though. Is it the hard requirement to hide the sidebar by default?


Screen Shot 2023-01-05 at 6 28 13 PM

**Comment by ericl** [comment]
I guess it's fine.

**Review by richardliaw — APPROVED** [review]
looks good for a first pass!

**Comment by maxpumperla** [comment]
@c21 I think the sidebar behaviour can be considered a feature (we could still change it later). One of the major points here is to introduce another hierarchy. If a users drills down 3 levels to learn about our API, I don't see the issue with having granular info in the sidebar.

**Review by maxpumperla — APPROVED** [review]
lgtm

**Review by clarkzinzow — APPROVED** [review]
LGTM, great work, this is a huge improvement! 👏

**Pull request merged**